File: mobile/lib/models/course.dart
class Course {
  final int id;
  final String courseCode;
  final String courseTitle;
  final String lecturer;
  final String? faculty;
  final int creditHours;
  final int maxCapacity;
  final int currentCapacity;
  final int enrolledCount;
  final String semester;

  Course({
    required this.id,
    required this.courseCode,
    required this.courseTitle,
    required this.lecturer,
    this.faculty,
    required this.creditHours,
    required this.maxCapacity,
    required this.currentCapacity,
    required this.enrolledCount,
    required this.semester,
  });

  factory Course.fromJson(Map<String, dynamic> json) {
    return Course(
      id: json['id'] as int? ?? 0,
      courseCode: json['courseCode'] as String? ?? '',
      courseTitle: json['courseTitle'] as String? ?? '',
      lecturer: json['lecturer'] as String? ?? '',
      faculty: json['faculty'] as String?,
      creditHours: json['creditHours'] as int? ?? 3,
      maxCapacity: json['maxCapacity'] as int? ?? 30,
      currentCapacity: json['currentCapacity'] as int? ?? 0,
      enrolledCount: json['enrolledCount'] as int? ?? 0,
      semester: json['semester']?.toString() ?? '1',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'courseCode': courseCode,
      'courseTitle': courseTitle,
      'lecturer': lecturer,
      'faculty': faculty,
      'creditHours': creditHours,
      'maxCapacity': maxCapacity,
      'currentCapacity': currentCapacity,
      'enrolledCount': enrolledCount,
      'semester': semester,
    };
  }
}

File: enrolment-service/src/main/java/smartcampus/enrolment/model/Course.java
package smartcampus.enrolment.model;

import jakarta.persistence.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "courses")
public class Course {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "course_code", nullable = false, unique = true, length = 20)
    private String courseCode;

    @Column(name = "course_title", nullable = false, length = 150)
    private String courseTitle;

    @Column(length = 100)
    private String lecturer;

    @Column(length = 50)
    private String faculty;

    @Column(name = "credit_hours", nullable = false)
    private Integer creditHours = 3;

    /**
     * Shared mutable state protected by ReentrantLock in EnrolmentService (R5).
     */
    @Column(name = "current_capacity", nullable = false)
    private Integer currentCapacity = 0;

    @Column(name = "max_capacity", nullable = false)
    private Integer maxCapacity = 30;

    @Column(length = 20)
    private String semester;

    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    public void onCreate() { this.createdAt = LocalDateTime.now(); }

    public Course() {}

    public Course(String courseCode, String courseTitle, String lecturer, String faculty,
                  Integer creditHours, Integer maxCapacity, String semester) {
        this.courseCode = courseCode; this.courseTitle = courseTitle;
        this.lecturer = lecturer; this.faculty = faculty;
        this.creditHours = creditHours; this.maxCapacity = maxCapacity;
        this.semester = semester;
    }

    public boolean isFull() { return this.currentCapacity >= this.maxCapacity; }
    public int getRemainingSeats() { return this.maxCapacity - this.currentCapacity; }

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getCourseCode() { return courseCode; }
    public void setCourseCode(String courseCode) { this.courseCode = courseCode; }
    public String getCourseTitle() { return courseTitle; }
    public void setCourseTitle(String courseTitle) { this.courseTitle = courseTitle; }
    public String getLecturer() { return lecturer; }
    public void setLecturer(String lecturer) { this.lecturer = lecturer; }
    public String getFaculty() { return faculty; }
    public void setFaculty(String faculty) { this.faculty = faculty; }
    public Integer getCreditHours() { return creditHours; }
    public void setCreditHours(Integer creditHours) { this.creditHours = creditHours; }
    public Integer getCurrentCapacity() { return currentCapacity; }
    public void setCurrentCapacity(Integer currentCapacity) { this.currentCapacity = currentCapacity; }
    public Integer getMaxCapacity() { return maxCapacity; }
    public void setMaxCapacity(Integer maxCapacity) { this.maxCapacity = maxCapacity; }
    public String getSemester() { return semester; }
    public void setSemester(String semester) { this.semester = semester; }
    public LocalDateTime getCreatedAt() { return createdAt; }
}
